### PR TITLE
Ruby Marionette/Wires/W3C Support

### PR DESCRIPTION
**Pull request by titusfortner**
@p0deje - The goal of this PR is for user tests to work seamlessly with current versions of Firefox and Wires-only versions of Firefox. If their version of Firefox is incompatible with the current extension-based implementation, then this will raise an error for a user to download wires and put it in their PATH.

To test this out, you need to download wires: https://github.com/jgraham/wires/releases & Firefox Nightly. If you are on Linux and have multiple copies of firefox (Nightly & Release), you need to specify the path to the nightly firefox-bin file in ENV['FIREFOX_NIGHTLY_PATH'].

I've run the Ruby specs for this PR on all supported browsers in Windows, Linux & Mac. All tests pass with this code, or are guarded against with specific notes as to what is preventing it passing. This is definitely not the best abstracted or most elegant solution, but it works and is completely backward compatible.

Feedback is appreciated, otherwise I'll merge it in and continue to clean it up as more of the W3C protocol is implemented.

### CODE REVIEW

**Comment by p0deje** [comment]
That's a lot of changes, please give me some time to go over both styles and logic.

File: rb/lib/selenium/webdriver/common/driver.rb [hunk 1]
```diff
@@ -47,7 +47,14 @@ def for(browser, opts = {})
 
           bridge = case browser
                    when :firefox, :ff
-                     Firefox::Bridge.new(opts)
+                     if  Remote::W3CCapabilities.w3c?(opts)
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/driver.rb]: extra space

File: rb/lib/selenium/webdriver/common/driver.rb [hunk 2]
```diff
@@ -47,7 +47,14 @@ def for(browser, opts = {})
...
+                       Firefox::W3CBridge.new(opts)
+                     else
+                       Firefox::Bridge.new(opts)
+                     end
+                   when :firefox_nightly, :nightly
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/driver.rb]: maybe it's better to avoid `:nightly` since it's too common?

File: rb/lib/selenium/webdriver/common/error.rb [hunk 3]
```diff
@@ -216,8 +216,12 @@ class MoveTargetOutOfBoundsError < WebDriverError; end     # 34
       class << self
         def for_code(code)
           return if [nil, 0].include? code
+          return Errors[code -1] if code.is_a? Fixnum
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/error.rb]: space after `-`

File: rb/lib/selenium/webdriver/common/error.rb [hunk 4]
```diff
@@ -216,8 +216,12 @@ class MoveTargetOutOfBoundsError < WebDriverError; end     # 34
...
 
-          Errors[code - 1] || WebDriverError
+          klass_name = code.split(' ').map(&:capitalize!).join
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/error.rb]: use non-destructive `:capitalize` as bang-version might return `nil`.

File: rb/lib/selenium/webdriver/common/error.rb [hunk 5]
```diff
@@ -216,8 +216,12 @@ class MoveTargetOutOfBoundsError < WebDriverError; end     # 34
...
+          Error.const_get("#{klass_name}Error")
+        rescue NameError
+          return WebDriverError
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/error.rb]: unnecessary return

File: rb/lib/selenium/webdriver/common/w3c_error.rb [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+      NoAlertPresentError       = NoSuchAlertError
+      ScriptTimeOutError        = ScriptTimeoutError
+
+      ObsoleteElementError      = StaleElementReferenceError
+      UnhandledError            = UnknownError
+      UnexpectedJavascriptError = JavascriptError
+      NoAlertOpenError          = NoAlertPresentError
+      ElementNotDisplayedError  = ElementNotVisibleError
+
+
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/w3c_error.rb]: remove empty line

File: rb/lib/selenium/webdriver/firefox/binary.rb [hunk 7]
```diff
@@ -149,6 +149,33 @@ def path
...
           end
 
+          def version
+            @version ||= case Platform.os
+                         when :macosx
+                           `#{path} -v`.strip[/[^\s]*$/][/^\d+/].to_i
+                         when :windows
+                           `\"#{path}\" -v | more`.strip[/[^\s]*$/][/^\d+/].to_i
+                         when :linux
+                          `firefox -v`.strip[/[^\s]*$/][/^\d+/].to_i
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/firefox/binary.rb]: indentation

File: rb/lib/selenium/webdriver/firefox/binary.rb [hunk 8]
```diff
@@ -149,6 +149,33 @@ def path
...
+                         else
+                             0
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/firefox/binary.rb]: indentation

File: rb/lib/selenium/webdriver/firefox/binary.rb [hunk 9]
```diff
@@ -149,6 +149,33 @@ def path
...
+          end
+
+          def nightly_path
+            nightly_path = case Platform.os
+                           when :macosx
+                             "/Applications/FirefoxNightly.app/Contents/MacOS/firefox-bin"
+                           when :windows
+                             "#{Platform.find_in_program_files("\\Nightly\\firefox.exe").gsub('m Files (x86)', '~2')}"
+                           when :linux
+                            ENV['FIREFOX_NIGHTLY_PATH']
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/firefox/binary.rb]: indentation

File: rb/lib/selenium/webdriver/firefox/w3c_bridge.rb [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        def quit
+          super
+        ensure
+          @service.stop if @service
+        end
+
+        private
+
+        def create_capabilities(opts)
+          caps               = opts.delete(:desired_capabilities) { Remote::W3CCapabilities.firefox }
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/firefox/w3c_bridge.rb]: extra spaces

File: rb/lib/selenium/webdriver/firefox/w3c_bridge.rb [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+          args = []
+
+          if opts.has_key?(:service_log_path)
+            args << "--log-path=#{opts.delete(:service_log_path)}"
+          end
+
+          args
+        end
+
+      end # Bridge
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/firefox/w3c_bridge.rb]: W3CBridge

File: rb/lib/selenium/webdriver/remote/response.rb [hunk 12]
```diff
@@ -104,11 +103,11 @@ def add_backtrace(ex)
         end
 
         def status
-          @payload['status']
+          @payload['status'] || @payload['error']
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/response.rb]: Wonder why this change is needed

File: rb/lib/selenium/webdriver/remote/w3c_bridge.rb [hunk 13]
```diff
@@ -0,0 +1,676 @@
...
+        include BridgeHelper
+
+        COMMANDS = {}
+
+        #
+        # Defines a wrapper method for a command, which ultimately calls #execute.
+        #
+        # @param name [Symbol]
+        #   name of the resulting method
+        # @param url [String]
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_bridge.rb]: Swap url and verb params

File: rb/lib/selenium/webdriver/remote/w3c_bridge.rb [hunk 14]
```diff
@@ -0,0 +1,676 @@
...
+        def getVisible
+          execute :getVisible
+        end
+
+        def setVisible(bool)
+          execute :setVisible, {}, bool
+        end
+
+        def switchToWindow(name)
+            execute :switchToWindow, {}, :handle => name
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_bridge.rb]: indentation

File: rb/lib/selenium/webdriver/remote/w3c_bridge.rb [hunk 15]
```diff
@@ -0,0 +1,676 @@
...
+
+          puts "-> #{verb.to_s.upcase} #{path}" if $DEBUG
+          http.call verb, path, command_hash
+        end
+
+        def escaper
+          @escaper ||= defined?(URI::Parser) ? URI::Parser.new : URI
+        end
+
+      end # Bridge
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_bridge.rb]: W3CBridge

File: rb/lib/selenium/webdriver/remote/w3c_capabilities.rb [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+  module WebDriver
+    module Remote
+      #
+      # Specification of the desired and/or actual capabilities of the browser that the
+      # server is being asked to create.
+      #
+      class W3CCapabilities
+
+        DEFAULTS = {
+            :browser_name => '',
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: indentation

File: rb/lib/selenium/webdriver/remote/w3c_capabilities.rb [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+
+        #
+        # Convenience methods for the common choices.
+        #
+
+        class << self
+
+          def edge(opts = {})
+            new({
+                    :browser_name => "MicrosoftEdge",
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: indentation

File: rb/lib/selenium/webdriver/remote/w3c_capabilities.rb [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+                    :platform => :windows,
+                }.merge(opts))
+          end
+
+          def firefox(opts = {})
+            new({
+                    :browser_name => "firefox"
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: indentation

File: rb/lib/selenium/webdriver/remote/w3c_capabilities.rb [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+                }.merge(opts))
+          end
+
+          def w3c?(opts={})
+            return true if opts[:browser_name] == :nightly
+            intersect = Remote::W3CCapabilities::DEFAULTS.keys & opts.keys
+            common = [:takes_screenshot, :proxy]
+            !(intersect - common).empty? || Firefox::Binary.version > 43
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: this method asks for splitting into smaller and more descriptive ones

File: rb/lib/selenium/webdriver/remote/w3c_capabilities.rb [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+            caps.proxy = Proxy.json_create(data['proxy']) if data.has_key?('proxy')
+
+            # any remaining pairs will be added as is, with no conversion
+            caps.merge!(data)
+
+            caps
+          end
+        end
+
+        # @option :browser_name           [String] required browser name
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: alignment of all these params

Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: also, YARD asks for `@param ` the `@option` is attached to

File: rb/lib/selenium/webdriver/remote/w3c_capabilities.rb [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+        # @option :browser_version                [String] required browser version number
+        # @option :platform_name               [Symbol] one of :any, :win, :mac, or :x
+        # @option :platform_version     [String] required platform version number
+        # @option :accept_ssl_certs  [Boolean] does the driver accept SSL Cerfifications?
+        # @option :takes_screenshot       [Boolean] can this driver take screenshots?
+        # @option :takes_element_screenshot       [Boolean] can this driver take element screenshots?
+        # @option :proxy                  [Selenium::WebDriver::Proxy, Hash] proxy configuration
+        #
+        # @api public
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: is it really needed

File: rb/lib/selenium/webdriver/remote/w3c_capabilities.rb [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+          @capabilities
+        end
+
+        private
+
+        def camel_case(str)
+          str.gsub(/_([a-z])/) { $1.upcase }
+        end
+
+      end # Capabilities
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: W3CCapabilities

File: rb/lib/selenium/webdriver/remote/w3c_commands.rb [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+
+  #
+  # http://www.w3.org/TR/2015/WD-webdriver-20150918/#list-of-endpoints
+  #
+
+  #
+  # session handling
+  #
+
+  command :newSession,                             :post,    "session"
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/remote/w3c_commands.rb]: spaces :)

File: rb/spec/integration/selenium/webdriver/spec_support/test_environment.rb [hunk 24]
```diff
@@ -107,6 +108,10 @@ def native_events?
           @native_events ||= !!ENV['native']
         end
 
+        def w3c?(opt={})
```
Comment (p0deje) [inline rb/spec/integration/selenium/webdriver/spec_support/test_environment.rb]: spaces around `=`

**Comment by p0deje** [comment]
:+1: :+1: :+1: 

That is really huge and amazing amount of work, thanks for doing that! Regardless of all the tiny comments I've posted, I believe it makes a huge Ruby bindings release

File: rb/lib/selenium/webdriver/remote/response.rb (hunk 12 above)
Comment (titusfortner) [inline rb/lib/selenium/webdriver/remote/response.rb]: I'll verify this is per spec.

File: rb/lib/selenium/webdriver/remote/w3c_capabilities.rb (hunk 21 above)
Comment (titusfortner) [inline rb/lib/selenium/webdriver/remote/w3c_capabilities.rb]: I'm not sure, it is literally copied over from existing code - https://github.com/SeleniumHQ/selenium/blob/master/rb/lib/selenium/webdriver/remote/capabilities.rb#L193
I don't know how yard uses the public/private designations.

File: rb/lib/selenium/webdriver/remote/response.rb (hunk 12 above)
Comment (titusfortner) [inline rb/lib/selenium/webdriver/remote/response.rb]: Yeah, it's 'error' & 'message' - http://www.w3.org/TR/webdriver/#processing-model
There are several things in this Response class that are handled differently between implementations, but I don't see how to address that right now. Something to consider down the road.

**Comment by titusfortner** [comment]
Man, I really want to get this code up to >2.0 compliant so that Rubocop can do this stuff for me/us. :)
Thanks for catching all of that. I'll rebase the commits and merge it later today.
